src/handler/sign_tx.c

```diff
@@ -1,103 +1,423 @@
+        if (input_val>0) {
+            // we require input to have no data, since we are signing all data received (sighash_all must have no data)
+            THROW(TX_STATE_ERR);
         }
+        // reading input
+        G_context.tx_info.remaining_inputs--;
+        G_context.tx_info.buffer_len -= TX_INPUT_LEN;
+        G_context.tx_info.elem_type = ELEM_INPUT;
+        memmove(G_context.tx_info.buffer, G_context.tx_info.buffer + TX_INPUT_LEN, G_context.tx_info.buffer_len);
+    } else if (G_context.tx_info.current_output < G_context.tx_info.outputs_len) {
+        // read output
+        size_t output_len = parse_output(G_context.tx_info.buffer, G_context.tx_info.buffer_len, &G_context.tx_info.decoded_output);
+        G_context.tx_info.decoded_output.index = G_context.tx_info.current_output;
+        G_context.tx_info.elem_type = ELEM_OUTPUT;
+        // move buffer
+        G_context.tx_info.buffer_len -= output_len;
+        memmove(G_context.tx_info.buffer, G_context.tx_info.buffer + output_len, output_len); 
+        G_context.tx_info.current_output++;
+    } else {
+        // we reached the end of the data we should read, the buffer should be empty
+        // if the buffer is not empty something went wrong
+        if (G_context.tx_info.buffer_len > 0) THROW(TX_STATE_ERR);
+        THROW(TX_STATE_FINISHED);
+    }
 
-        return io_send_sw(SW_OK);
-    } else {  // parse transaction
-        if (G_context.req_type != CONFIRM_TRANSACTION) {
-            return io_send_sw(SW_BAD_STATE);
+    switch (G_context.tx_info.elem_type)
+    {
+    case ELEM_TOKEN_UID:
+        // not displaying token uid
+        break;
+    case ELEM_INPUT:
+        // not displaying inputs
+        break;
+    case ELEM_OUTPUT:
+        // check if this is the change output
+        if (
+            G_context.tx_info.has_change_output &&
+            G_context.tx_info.change_output_index == G_context.tx_info.decoded_output.index) {
+                // verify change
+                if (!verify_change_output(
+                    G_context.tx_info.decoded_output,
+                    G_context.tx_info.change_bip32_path.path,
+                    G_context.tx_info.change_bip32_path.length)) {
+                    THROW(TX_STATE_ERR);
+                }
+        } else {
+            THROW(TX_STATE_READY);
         }
+        break;
+    }
+}
 
-        if (more) {  // more APDUs with transaction part
-            if (G_context.tx_info.raw_tx_len + cdata->size > MAX_TRANSACTION_LEN &&  //
-                !buffer_move(cdata,
-                             G_context.tx_info.raw_tx + G_context.tx_info.raw_tx_len,
-                             cdata->size)) {
-                return io_send_sw(SW_WRONG_TX_LENGTH);
-            }
+tx_decoder_state_e decode_next_element() {
+    volatile tx_decoder_state_e result;
+    BEGIN_TRY {
+        TRY {
+            for (;;) _decode_next_element();
```

Comment:
I understood that `_decode_next_element()` will throw an exception when it wants to finish, right? Maybe we should change `_decode_next_element()` to return `1` when it should be called again and do 

```c
while(_decode_next_element()) {
    ;
}
```

Reply:
Done.